easycv/datasets/detection/data_sources/base.py

```diff
@@ -0,0 +1,171 @@
+            raise ValueError(
+                'Only one of `cache_on_the_fly` and `cache_at_init` can be True!'
+            )
+        source_iter = self.get_source_iterator()
+
+        process_fn = functools.partial(
+            build_sample,
+            parse_fn=parse_fn,
+            classes=self.CLASSES,
+            load_img=cache_at_init == True,
+        )
+        self.samples_list = self.build_samples(
+            source_iter, process_fn=process_fn)
+
+    @abstractmethod
+    def get_source_iterator():
+        """data list iterator, for multi-process read
+        """
+        raise NotImplementedError
+
+    def build_samples(self, iterable, process_fn):
+        samples_list = []
+        with Pool(processes=self.num_processes) as p:
+            with tqdm(total=len(iterable), desc='Scanning images') as pbar:
+                for _, result_dict in enumerate(
+                        p.imap_unordered(process_fn, iterable)):
+                    if result_dict:
+                        samples_list.append(result_dict)
+                    pbar.update()
+
+        return samples_list
+
+    def get_length(self):
+        return len(self.samples_list)
+
+    def __len__(self):
+        return self.get_length()
+
+    def get_ann_info(self, idx):
+        """
+        Get raw annotation info, include bounding boxes, labels and so on.
+        `bboxes` format is as [x1, y1, x2, y2] without normalization.
+        """
+        sample_info = self.samples_list[idx]
+
+        groundtruth_is_crowd = sample_info.get('groundtruth_is_crowd', None)
+        if groundtruth_is_crowd is None:
+            groundtruth_is_crowd = np.zeros_like(sample_info['gt_labels'])
+
+        annotations = {
+            'bboxes': sample_info['gt_bboxes'],
+            'labels': sample_info['gt_labels'],
+            'groundtruth_is_crowd': groundtruth_is_crowd
+        }
+
+        return annotations
+
+    def post_process_fn(self, result_dict):
+        if result_dict.get('img_fields', None) is None:
+            result_dict['img_fields'] = ['img']
+        if result_dict.get('bbox_fields', None) is None:
+            result_dict['bbox_fields'] = ['gt_bboxes']
+
+        return result_dict
+
+    def get_sample(self, idx):
+        result_dict = self.samples_list[idx]
+        load_success = True
+        try:
+            if result_dict.get('img', None) is None:
+                result_dict.update(load_image(result_dict['filename']))
+                if self.cache_at_init or self.cache_on_the_fly:
+                    self.samples_list[idx] = result_dict
+
+            result_dict = self.post_process_fn(result_dict)
+        except Exception as e:
+            logging.error(e)
+            load_success = False
+
+        if not load_success:
+            logging.warning(
+                'Something wrong with current sample %s,Try load next sample...'
+                % result_dict.get('filename', ''))
+            result_dict = self.get_sample(idx + 1)
```

Comment:
this recursive call may result in infinite loop when dealing last exmaple  with exception, the index will go out of bounds, should consider data source length and do mod operation

Reply:
done! and add _max_retry_num to avoid looping all the time